lib/message_request.js

```diff
@@ -0,0 +1,175 @@
+  if (request) {
+    request.setOptions(options);
+  }
+
+  return request;
+};
+
+Request.buildPublish = function (scope, value) {
+  var request = new Request('publish', scope);
+  if (request) {
+    request.setAttr('value', value);
+  }
+
+  return request;
+};
+
+Request.buildPush = function (scope, resource, action, value) {
+  var request = new Request('push', scope);
+  if (request) {
+    request.setAttr('resource', resource);
+    request.setAttr('action', action);
+    request.setAttr('value', value);
+  }
+
+  return request;
+};
+
+Request.buildNameSync = function (scope, options) {
+  var request = new Request('nameSync', scope);
+  if (request) {
+    request.setOptions(options);
+  }
+
+  return request;
+};
+
+Request.buildSet = function (scope, value, key, userType) {
+  var request = new Request('set', scope);
+
+  if (request) {
+    request.setAttr('value', value);
+    request.setAttr('key', key);
+    request.setAttr('type', userType);
+  }
+
+  return request;
+};
+
+Request.buildSync = function (scope) {
+  return new Request('sync', scope);
+};
+
+Request.buildSubscribe = function (scope) {
+  return new Request('subscribe', scope);
+};
+
+Request.buildUnsubscribe = function (scope, options) {
+  return new Request('unsubscribe', scope);
+};
+
+// TO DO: config class should return the required data via getters
+Request.prototype.setAuthData = function (configuration) {
+  if (configuration && configuration.auth) {
+    this.setAttr('auth', configuration.auth);
+    this.setAttr('userId', configuration.userId);
+    this.setAttr('userType', configuration.userType);
+    this.setAttr('accountName', configuration.accountName);
+  }
+};
+
+Request.getAttr = function (message, attr) {
+  return (message && message[attr]);
+};
+
+// Instance methods
+
+Request.prototype.getMessage = function () {
+  return this.message;
+};
+
+Request.prototype.setOptions = function (options) {
+  if (options) {
+    this.message.options = options;
+    this.message.options.version = 2;
+  } else {
+    this.message.options = { version: 1 };
+  }
+
+  return this;
+};
+
+Request.prototype.isPresence = function () {
+  return this.type === 'presence';
+};
+
+Request.prototype.setAttr = function (keyName, keyValue) {
+  if (keyName) {
+    this.message[keyName] = keyValue;
+  }
+};
+
+Request.prototype.getAttr = function (keyName) {
+  if (keyName) {
+    return this.message[keyName];
+  }
+};
+
+Request.prototype.getVersion = function () {
+  return this.message.options.version;
+};
+
+// Private methods
+
+Request.prototype._isValid = function () {
+  if (!this.message.op || !this.message.to) {
+    return false;
+  }
+
+  var type = this._getType();
+  if (type) {
+    this.type = type;
+    return this._isValidType(type) && this._isValidOperation(type);
+  } else {
+    logger.error('missing type');
+  }
+  return false;
+};
+
+Request.prototype._isValidType = function () {
+  var types = Object.keys(opTable);
```

Comment:
I found inconsistent info on the IE browsers, both for Object.keys, and a proposed backfill, so I opted to extract the keys to a separate array and use indexOf.